rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -178,26 +179,62 @@ class DebruijnInterpreter[M[_], F[_]](
       else if (terms.size > 256) rand.splitShort(id.toShort)
       else rand.splitByte(id.toByte)
 
-    terms.zipWithIndex.toList.parTraverse_ {
-      case (term, id) =>
-        implicit val rand: Blake2b512Random = split(id)
-        term match {
-          case term: Send    => reportErrors(eval(term))
-          case term: Receive => reportErrors(eval(term))
-          case term: New     => reportErrors(eval(term))
-          case term: Match   => reportErrors(eval(term))
-          case term: Bundle  => reportErrors(eval(term))
-          case term: Expr =>
-            term.exprInstance match {
-              case e: EVarBody    => reportErrors(eval(e.value.v) >>= (eval(_)))
-              case e: EMethodBody => reportErrors(evalExprToPar(e) >>= (eval(_)))
-              case other          => BugFoundError(s"Undefined term: \n $other").raiseError[M, Unit]
-            }
-          case other => BugFoundError(s"Undefined term: \n $other").raiseError[M, Unit]
+    val indexedTerms = terms.zipWithIndex.toList
+    if (terms.size <= Reduce.parallelism) {
+      indexedTerms.parTraverse_ {
+        case (term, index) =>
+          val random = split(index)
+          eval(term)(env, random, sequenceNumber)
+      }
+    } else {
+      //TODO: Investigate if we can avoid the shuffling and manual parallelism limiting by tweaking:
+      // - the scheduler used
+      // - the monix ExecutionModel used
+      // Verify if using traverse-per-batch inside parTraverse means that less tasks are spawned at once
+      // (surmise: the task for the next element of a batch is not spawned until the previous one is processed)
+      // Track memory usage along with raw performance when investigating. Clean slate POC could be helpful here.
+
+      // we index terms before shuffling to have deterministic Blake2b512Random seeds for each term
+      val indexedTermsShuffled = Random.shuffle(indexedTerms)
+      val groups               = indexedTermsShuffled.groupBy(_._2 % Reduce.parallelism).values
+      groups.toList.parTraverse_ {
```

Comment:
@pyrocto and others - do you see an issue in here?

Notice we're retaining the random seed by indexing the terms prior to shuffling.

I think this is a faithful, if not slightly idealized, simulation of non-deterministic execution order of pars. It should also bring more fairness and bugs found due to a more uniform distribution of execution orders too.

Does anybody see issues with the approach?

Reply:
This looks fine to me.